python/ray/tune/ray_trial_executor.py

```diff
@@ -469,10 +469,37 @@ def save(self, trial, storage=Checkpoint.DISK):
         if storage == Checkpoint.MEMORY:
             trial._checkpoint.value = trial.runner.save_to_object.remote()
         else:
-            with warn_if_slow("save_to_disk"):
-                trial._checkpoint.value = ray.get(trial.runner.save.remote())
+            # Keeps only highest performing checkpoints
+            if trial.keep_checkpoints_num and trial.results_since_checkpoint_cnt > 0:
+                current_attr_mean = trial.results_since_checkpoint_sum / trial.results_since_checkpoint_cnt
+                if trial.compare_checkpoints(current_attr_mean):
+                    trial.best_checkpoint_attr_value = current_attr_mean
+                    self._checkpoint_and_erase(trial)
+                trial.results_since_checkpoint_sum, trial.results_since_checkpoint_cnt = 0, 0
+            else:
+                with warn_if_slow("save_to_disk"):
+                    trial._checkpoint.value = ray.get(trial.runner.save.remote())
+
         return trial._checkpoint.value
 
+    def _checkpoint_and_erase(self, trial):
+        """Checkpoints the model and erases old checkpoints
+            if needed.
+        Parameters
+        ----------
+            subdir string: either "" or "best"
+            trial : trial to save
+        """
+
+        with warn_if_slow("save_to_disk"):
+            trial._checkpoint.value = ray.get(trial.runner.save.remote())
+
+        if len(trial.history) >= trial.keep_checkpoints_num:
+            ray.get(trial.runner.delete_checkpoint.remote(trial.history[-1]))
+            trial.history.pop()
```

Comment:
If I understand this logic here, we keep a FIFO history of trials. However you only add trials to the queue if it exceeds the mean threshold since the last checkpoint?

Why not instead retain the exact top K trials?

Reply:
I can see the value of this compared to a normal top-K. I think it's still somewhat difficult to understand however.

How about this simplification: don't keep a running mean; just compare against the previous top result? RLlib already internally smoothes the trial so doing an additional average isn't helping that much. It can also lead to degenerate cases since you don't reset the averaging window.

Reply:
Okay, done! It's not taking average anymore. Just looking at the last_result and comparing to previous seen. Good enough for PR?